tsl/src/continuous_aggs/refresh.c

```diff
@@ -0,0 +1,381 @@
+			break;
+		case TIMESTAMPTZOID:
+		case TIMESTAMPOID:
+			maxrange.start = MIN_TIMESTAMP;
+			maxrange.end = (END_TIMESTAMP - TS_EPOCH_DIFF_MICROSECONDS) - 1;
+			break;
+		case INT8OID:
+		case INT4OID:
+		case INT2OID:
+			/* There is no "date"/"time" range for integers so they can take
+			 * any value. */
+			maxrange.start = PG_INT64_MIN;
+			maxrange.end = PG_INT64_MAX;
+			break;
+		default:
+			elog(ERROR, "unrecognized time type %d", timetype);
+			break;
+	}
+
+	maxrange.start = ts_time_value_to_internal(Int64GetDatum(maxrange.start), timetype);
+	maxrange.end = ts_time_value_to_internal(Int64GetDatum(maxrange.end), timetype);
+
+	return maxrange;
+}
+
+/*
+ * Compute the largest possible bucketed window given the time type and
+ * internal restrictions.
+ *
+ * The largest bucketed window is governed by restrictions set by the type and
+ * internal, TimescaleDB-specific legacy details (see get_max_window above for
+ * further explanation).
+ */
+static InternalTimeRange
+get_largest_bucketed_window(Oid timetype, int64 bucket_width)
+{
+	InternalTimeRange maxwindow = get_max_window(timetype);
+	InternalTimeRange maxbuckets;
+
+	/* For the MIN value, the corresponding bucket either falls on the exact
+	 * MIN or it will be below it. Therefore, we add (bucket_width - 1) to
+	 * move to the next bucket to be within the allowed range. */
+	maxwindow.start = maxwindow.start + bucket_width - 1;
+	maxbuckets.start = ts_time_bucket_by_type(bucket_width, maxwindow.start, timetype);
+	maxbuckets.end = ts_time_bucket_by_type(bucket_width, maxwindow.end, timetype);
+
+	return maxbuckets;
+}
+
+/*
+ * Adjust the refresh window to align with buckets in an inclusive manner.
+ *
+ * It is OK to refresh more than the given refresh window, but not less. Since
+ * we can only refresh along bucket boundaries, we need to adjust the refresh
+ * window to be inclusive in both ends to be able to refresh the given
+ * region. For example, if the dotted region below is the original window, the
+ * adjusted refresh window includes all four buckets shown.
+ *
+ * | ....|.....|..   |
+ */
+static InternalTimeRange
+compute_bucketed_refresh_window(const InternalTimeRange *refresh_window, int64 bucket_width)
+{
+	InternalTimeRange result = *refresh_window;
+	InternalTimeRange largest_bucketed_window =
+		get_largest_bucketed_window(refresh_window->type, bucket_width);
+
+	if (result.start <= largest_bucketed_window.start)
+		result.start = largest_bucketed_window.start;
+	else
+		result.start = ts_time_bucket_by_type(bucket_width, result.start, result.type);
+
+	if (result.end >= largest_bucketed_window.end)
+		result.end = largest_bucketed_window.end;
+	else
+	{
+		/* We get the time value for the start of the bucket, so need to add
+		 * bucket_width to get the end of it */
+		result.end = ts_time_bucket_by_type(bucket_width, result.end, result.type);
+		result.end = result.end + bucket_width;
```

Comment:
if the value is at a bucket edge, you are moving up by a bucket here.
say bucket_width = 10, result.end = 20, then final result.end = 30

Reply:
The edge case was already checked for in the if clause, so we must be below the edge, which means that the bucket start must be at least one bucket down from the edge and this would just bring it at most up to the edge again, which is OK.